File: ddd-application/src/main/java/victor/training/ddd/agile/entity/BacklogItem.java
package victor.training.ddd.agile.entity;

import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import javax.persistence.*;
import javax.validation.constraints.NotNull;

import static javax.persistence.EnumType.STRING;

@Getter
@Setter
@NoArgsConstructor
@Entity
public class BacklogItem {
   @Id
   @GeneratedValue
   private Long id;
   @ManyToOne
   private Product product;

@NotNull
   private String title;
   private String description;

  public void complete() {
      if (getStatus() != Status.STARTED) {
          throw new IllegalStateException("Cannot complete an Item before starting it");
      }
      setStatus(Status.DONE);
  }

  public void start() {
//    if (sprint.getStatus() != Sprint.Status.STARTED) {
//      throw new IllegalStateException();
//    }
      if (getStatus() != Status.CREATED) {
          throw new IllegalStateException("Item already started");
      }
      setStatus(Status.STARTED);
  }

  public boolean isDone() {
    return status == Status.DONE;
  }

  public enum Status {
      CREATED,
      STARTED,
      DONE
   }
   @Enumerated(STRING)
   private Status status = Status.CREATED;

   @ManyToOne
   private Sprint sprint; // ⚠ not NULL when assigned to a sprint
   private Integer fpEstimation; // ⚠ not NULL when assigned to a sprint

   private int hoursConsumed;

   @Version
   private Long version;

   public void addHours(int hours) {
      hoursConsumed += hours;
   }

}

File: ddd-application/src/main/java/victor/training/ddd/agile/entity/BacklogItem.kt
package victor.training.ddd.agile.entity

import javax.persistence.*
import javax.validation.constraints.NotNull

// Child entity of the "Sprint" Aggregate.
@Entity
class BacklogItem(
    @ManyToOne val product: Product,
    val title: @NotNull String?,
    val description: String,
    @Version val version: Long? = null,
    @GeneratedValue @Id val id: Long? = null,
    @Enumerated(EnumType.STRING)
    var status: ItemStatus = ItemStatus.CREATED,
    @ManyToOne
    var sprint: Sprint? = null, // ⚠ not NULL when assigned to a sprint
    var fpEstimation: Int? = null, // ⚠ not NULL when assigned to a sprint
    private var hoursConsumed: Int = 0
) {
    enum class ItemStatus {
        CREATED, STARTED, DONE
    }

    fun start() {
        check(status == ItemStatus.CREATED) { "Item already started" }
        status = ItemStatus.STARTED
    }

    fun complete() {
        check(status == ItemStatus.STARTED) { "Cannot complete an Item before starting it" }
        status = ItemStatus.DONE
    }

    fun hoursConsumed() = hoursConsumed
    fun logHours(hours: Int) {
        require(hours >= 0) {"cannot substract hours"}
        check(status === ItemStatus.STARTED) { "Item not started" }
        hoursConsumed += hours
    }

    fun isDone() = status === ItemStatus.DONE
}